qiling/os/posix/syscall/epoll.py

```diff
@@ -0,0 +1,251 @@
+from qiling.os.filestruct import ql_file
+import select
+from ctypes import *
+from qiling.os import struct
+import struct
+from qiling.os.filestruct import PersistentQlFile
+from qiling.extensions import pipe
+import sys
+
+class QlEpollObj:
+    def __init__(self, epoll_object):
+        self._epoll_object = epoll_object
+        self._fds = {}  # key: fd, value: eventmask
+        # keep track of which fds have what eventmasks,
+        # since this isn't directly supported in select.epoll
+
+    @property
+    def get_epoll_instance(self):
+        return self._epoll_object
+
+    @property
+    def get_eventmask(self, fd):
+        return self._fds[fd]
+
+    @property
+    def get_fds(self):
+        if len(self._fds.keys()) == 0:
+            return []
+        return list(self._fds.keys())
+
+    def set_eventmask(self, fd, newmask):
+        # the mask for an FD shouldn't ever be undefined
+        # as it is set whenever an FD is added for a QlEpollObj instance
+        newmask = self.get_eventmask() | newmask  # or with new eventmask value
+        self._epoll_object.modify(fd, newmask)
+
+    def monitor_fd(self, fd, eventmask):
+        self._epoll_object.register(
+            fd, eventmask
+        )  # tell the epoll object to watch the fd arg, looking for events matching the eventmask
+        self._fds[fd] = eventmask
+
+    def delist_fd(self, fd):
+        self._fds.pop(fd)
+        self._epoll_object.unregister(fd)
+
+    def close(self):
+        self.get_epoll_instance.close()
+
+    def is_present(self, fd):
+        if fd not in self.get_fds:
+            return 0
+        return 1
+
+'''
+Recursively checks each epoll instance's 'watched'
+fds for an instance of epoll being watched.
+If a chain of over 5 levels is detected, return
+1, which will return ELOOP in ql_syscall_epoll_wait
+'''
+def check_epoll_depth(ql_fd_list, epolls_list, depth):
+    if depth == 7:
+        return 1
+    new_epolls_list = []
+    flag = 0
+    for ent in list(epolls_list):
+        watched = ent.get_fds
+        for w in watched:
+            if isinstance(ql_fd_list[w], QlEpollObj):
+                flag = 1
+                new_epolls_list.append(ql_fd_list[w])
+        if flag:
+            check_epoll_depth(ql_fd_list, new_epolls_list, depth + 1)
+    return 0
+
+'''
+Modify an existing epoll
+man 7 epoll for more details
+'''
+def ql_syscall_epoll_ctl(ql: Qiling, epfd: int, op: int, fd: int, event: POINTER):
+    # Basic sanity checks first
+    if event != 0:
+        ql_event = ql.unpack32(ql.mem.read(event, 4))  # events list is uint32_t
+    else:
+        ql_event = (
+            0  # event can be null, for example, when deleting a fd from interest list
+        )
+    ql_op = ""
+    epoll_obj = -1
+    try:
+        epoll_parent_obj = ql.os.fd[epfd]
+        epoll_obj = epoll_parent_obj.get_epoll_instance
+    except KeyError as k:
+        ql.log.debug("Unable to grab epoll object, something wrong with ql.os.fd!")
```

Comment:
We should not warn about emulated software errors (i.e. when the emulated code calls a function with illegal parameters), only about errors or unexpected scenarios of Qiling itself (e.g. something was supposed to be found, but we can't find it).

If this is the former, we should remove the log message and stick with the returned error value.